Sources/Mimus/Comparator.swift

```diff
@@ -0,0 +1,73 @@
+internal class MimusComparator {
+    internal struct ComparisonResult {
+        struct MismatchedComparison {
+            let argumentIndex: Int             
+            let expected: Matcher?
+            let actual: Any?
+        }
+
+        let matching: Bool
+
+        let mismatchedComparisons: [MismatchedComparison]
+
+        init(matching: Bool) {
+            self.matching = matching
+            self.mismatchedComparisons = []
+        }
+
+        init(matching: Bool, mismatchedComparisons: [MismatchedComparison]) {
+            self.matching = matching
+            self.mismatchedComparisons = mismatchedComparisons
+        }
+    }
+
+    func match(expected: Arguments, actual: [Any?]?) -> ComparisonResult {
+        switch expected {
+        case .any:
+            return ComparisonResult(matching: true)
+        case .none:
+            return ComparisonResult(matching: actual == nil)
+        case let .actual(arguments):
+            return match(expected: arguments, actual: actual)
+        }
+    }
+
+    private func match(expected: [Matcher?], actual: [Any?]?) -> ComparisonResult {
+        guard let actualArguments = actual else {
+            return ComparisonResult(matching: false)
+        }
+
+        if expected.count != actualArguments.count {
+            return ComparisonResult(matching: false)
+        }
+
+        return match(expectedArguments: expected, actualArguments: actualArguments)
+    }
+
+    private func match(expectedArguments: [Matcher?], actualArguments: [Any?]) -> ComparisonResult {
```

Comment:
So technically actual arguments should never be of a `Matcher` type since these are arguments that were recorded in actual calls. 

We could add a test to see what happens when someone actually passes a matcher in actual arguments but I', not sure if it's worth it - we already have a hefty amount of tests for this in `ComparatorTests` and this sounds like testing a horrible misconfiguration 😅 (Also it'll just work since the matcher just won't match against itself or other matcher type 🤷).

Reply:
👍